### ISSUE TITLE
ZnapZend.pm respect system signals (SIGINT, SIGTERM, SIGHUP) only when executing a plan

### ISSUE THREAD

**Issue by chriseqipe** [oetiker/znapzend#168]
On debian wheezy, when sending SIGTERM or SIGINT to the perl process, znapsend.pm only checks for signals on the next run. If for example the plan takes a snapshot every 10 minutes, it takes maximum 10 minutes to end the process.  Maybe you should start a faster worker listening every minute to system signals.

**Comment by oetiker** [comment]
What would you like znapzend do when it receives a signal ? abort in mid operation ? for that you can send it a SIGKILL.

**Comment by martin-rueegg** [comment]
Thank you, @chriseqipe for pointing out this issue. I've come across it when writing an init.d script and wanted to _cleanly_ shutdown `znapzend`, which is probably what you wanted as well. My config was set to take one snapshot per hour, and it was quite frustrating that `znapzend` did not react on any signal (SIGTERM, SIGINT, SIGHUP), and the process could only be SIGKILLed (which is not performed by the script, but by the interpreter). After spending a day or so looking for the reason, i came across your issue and it helped me to track the cause.
 
unfortunately, @oetiker did not understand your post (of course, we _don't_ want to use SIGKILL, unless absolutely required!). so I'll try to explain it in other words:

The following is a diagram of the activity and events over time:
```
────> time ────>────>────>─────>─────>─────>────>────>──────>

signal─>─┬───────────────────────┬────────────────────┐
 ┌───────┴──┐                  ┌─┴─┐                ┌─┴─┐
[#######@@@@@]---------------*[@@@@@]-------------*[@@@@@]...
 ╘══╤══╛╘═╤═╛ ╘══════╤══════╛│ ╘═╤═╛ ╘═════╤═════╛│ ╘═╤═╛
   (1)   (2)        (3)     (4) (2)       (3)    (4) (2)
    │     │          │       │
    │     │          │       └─> (4) timer fires
    │     │          └─────────> (3) wait for next timer to fire
    │     │                          (-> which is not _sleep_!)
    │     └────────────────────> (2) Worker running
    └──────────────────────────> (1) startup
```
Signals will only be received and handled by `znapzend` during time of activity (between the `[brackets]`), as the program has no running code during the times where only timers are scheduled (`------`).

As such, if you have a schedule for taking one snapshot per hour, you'd have to wait up to one hour (when the next activity happens) to either reload the config or stop the daemon! Which leaves the SIGKILL variant to be the _only_ (as opposed to _last_) resort ...

This is not a problem of `znapzend` per se, as it is the Mojo::IOLoop (or rather the syscall to the timer functionality) that actually prevents the daemon from receiving those signals during the waiting periods.

It might be, that this issue should also be addressed by [Mojo:IOLoop](http://search.cpan.org/~sri/Mojolicious-6.47/lib/Mojo/IOLoop.pm), but I found the following two workarounds:

1. create a [recurring](http://search.cpan.org/~sri/Mojolicious-6.47/lib/Mojo/IOLoop.pm#recurring) timer to a function that will just `sleep`. however, this causes a lot of 'activity' as the interval needs to be small in order not to cause a leak after receiving a SIGHUP.
1.  the other solution is to [fork](http://search.cpan.org/dist/Mojo-IOLoop-ForkCall/lib/Mojo/IOLoop/ForkCall.pm#fork_call) such a function and respawn it only after a SIGHUP.

I considered the second approach to be the easier and will create a PR for the fix.

**Comment by oetiker** [comment]
hmmm ... I tried to verify ... so I wrote this little test program. It prints INTERVAL every 5 seconds and 'got HUP' whenever it receives a signal.

As far as I can see, the signals get received immediately ... several between intervals if sent.

```perl
use Mojo::IOLoop; 
$SIG{HUP} = sub { warn "got HUP\n" };
Mojo::IOLoop->recurring(5 => sub { warn "INTERVAL\n" });
Mojo::IOLoop->start;
```

**Comment by martin-rueegg** [comment]
I've tested your script (just added `print "$$\n";`), and in fact, it's the same issue

```
# perl k.pl
15150
INTERVAL
INTERVAL
got HUP
INTERVAL
got HUP
INTERVAL
got HUP
INTERVAL
got HUP
INTERVAL
INTERVAL
^C
#
```
while in an other console i sent SIGHUPs:
```
# kill -HUP 15150
# kill -HUP 15150
# kill -HUP 15150
# kill -HUP 15150
# kill -HUP 15150
# kill -HUP 15150
# kill -HUP 15150
# kill -HUP 15150
```

as expected, only on signal is received per interval.

```
# uname -a
Linux jackal 3.19.0-41-generic #46~14.04.2-Ubuntu SMP Tue Dec 8 17:46:10 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
# perl -V
Summary of my perl5 (revision 5 version 18 subversion 2) configuration:

  Platform:
    osname=linux, osvers=3.2.0-58-generic, archname=x86_64-linux-gnu-thread-multi
    uname='linux brownie 3.2.0-58-generic #88-ubuntu smp tue dec 3 17:37:58 utc 2013 x86_64 x86_64 x86_64 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -D_FORTIFY_SOURCE=2 -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -Dldflags= -Wl,-Bsymbolic-functions -Wl,-z,relro -Dlddlflags=-shared -Wl,-Bsymbolic-functions -Wl,-z,relro -Dcccdlflags=-fPIC -Darchname=x86_64-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.18 -Darchlib=/usr/lib/perl/5.18 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.18.2 -Dsitearch=/usr/local/lib/perl/5.18.2 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Duse64bitint -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -Ui_libutil -Uversiononly -DDEBUGGING=-g -Doptimize=-O2 -Duseshrplib -Dlibperl=libperl.so.5.18.2 -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fstack-protector -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fstack-protector -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.8.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=, so=so, useshrplib=true, libperl=libperl.so.5.18.2
    gnulibc_version='2.19'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib -fstack-protector'


Characteristics of this binary (from libperl):
  Compile-time options: HAS_TIMES MULTIPLICITY PERLIO_LAYERS
                        PERL_DONT_CREATE_GVSV
                        PERL_HASH_FUNC_ONE_AT_A_TIME_HARD
                        PERL_IMPLICIT_CONTEXT PERL_MALLOC_WRAP
                        PERL_PRESERVE_IVUV PERL_SAWAMPERSAND USE_64_BIT_ALL
                        USE_64_BIT_INT USE_ITHREADS USE_LARGE_FILES
                        USE_LOCALE USE_LOCALE_COLLATE USE_LOCALE_CTYPE
                        USE_LOCALE_NUMERIC USE_PERLIO USE_PERL_ATOF
                        USE_REENTRANT_API
  Locally applied patches:
        DEBPKG:debian/cpan_definstalldirs - Provide a sensible INSTALLDIRS default for modules installed from CPAN.
        DEBPKG:debian/db_file_ver - http://bugs.debian.org/340047 Remove overly restrictive DB_File version check.
        DEBPKG:debian/doc_info - Replace generic man(1) instructions with Debian-specific information.
        DEBPKG:debian/enc2xs_inc - http://bugs.debian.org/290336 Tweak enc2xs to follow symlinks and ignore missing @INC directories.
        DEBPKG:debian/errno_ver - http://bugs.debian.org/343351 Remove Errno version check due to upgrade problems with long-running processes.
        DEBPKG:debian/libperl_embed_doc - http://bugs.debian.org/186778 Note that libperl-dev package is required for embedded linking
        DEBPKG:fixes/respect_umask - Respect umask during installation
        DEBPKG:debian/writable_site_dirs - Set umask approproately for site install directories
        DEBPKG:debian/extutils_set_libperl_path - EU:MM: Set location of libperl.a to /usr/lib
        DEBPKG:debian/no_packlist_perllocal - Don't install .packlist or perllocal.pod for perl or vendor
        DEBPKG:debian/prefix_changes - Fiddle with *PREFIX and variables written to the makefile
        DEBPKG:debian/fakeroot - Postpone LD_LIBRARY_PATH evaluation to the binary targets.
        DEBPKG:debian/instmodsh_doc - Debian policy doesn't install .packlist files for core or vendor.
        DEBPKG:debian/ld_run_path - Remove standard libs from LD_RUN_PATH as per Debian policy.
        DEBPKG:debian/libnet_config_path - Set location of libnet.cfg to /etc/perl/Net as /usr may not be writable.
        DEBPKG:debian/mod_paths - Tweak @INC ordering for Debian
        DEBPKG:debian/module_build_man_extensions - http://bugs.debian.org/479460 Adjust Module::Build manual page extensions for the Debian Perl policy
        DEBPKG:debian/prune_libs - http://bugs.debian.org/128355 Prune the list of libraries wanted to what we actually need.
        DEBPKG:fixes/net_smtp_docs - [rt.cpan.org #36038] http://bugs.debian.org/100195 Document the Net::SMTP 'Port' option
        DEBPKG:debian/perlivp - http://bugs.debian.org/510895 Make perlivp skip include directories in /usr/local
        DEBPKG:debian/cpanplus_definstalldirs - http://bugs.debian.org/533707 Configure CPANPLUS to use the site directories by default.
        DEBPKG:debian/cpanplus_config_path - Save local versions of CPANPLUS::Config::System into /etc/perl.
        DEBPKG:debian/deprecate-with-apt - http://bugs.debian.org/702096 Point users to Debian packages of deprecated core modules
        DEBPKG:debian/squelch-locale-warnings - http://bugs.debian.org/508764 Squelch locale warnings in Debian package maintainer scripts
        DEBPKG:debian/skip-upstream-git-tests - Skip tests specific to the upstream Git repository
        DEBPKG:debian/patchlevel - http://bugs.debian.org/567489 List packaged patches for 5.18.2-2ubuntu1 in patchlevel.h
        DEBPKG:debian/skip-kfreebsd-crash - http://bugs.debian.org/628493 [perl #96272] Skip a crashing test case in t/op/threads.t on GNU/kFreeBSD
        DEBPKG:fixes/document_makemaker_ccflags - http://bugs.debian.org/628522 [rt.cpan.org #68613] Document that CCFLAGS should include $Config{ccflags}
        DEBPKG:debian/find_html2text - http://bugs.debian.org/640479 Configure CPAN::Distribution with correct name of html2text
        DEBPKG:debian/hurd_test_skip_stack - http://bugs.debian.org/650175 Disable failing GNU/Hurd tests dist/threads/t/stack.t
        DEBPKG:fixes/manpage_name_Test-Harness - http://bugs.debian.org/650451 [rt.cpan.org #73399] cpan/Test-Harness: add NAME headings in modules with POD
        DEBPKG:debian/makemaker-pasthru - http://bugs.debian.org/660195 [rt.cpan.org #28632] Make EU::MM pass LD through to recursive Makefile.PL invocations
        DEBPKG:debian/perl5db-x-terminal-emulator.patch - http://bugs.debian.org/668490 Invoke x-terminal-emulator rather than xterm in perl5db.pl
        DEBPKG:debian/cpan-missing-site-dirs - http://bugs.debian.org/688842 Fix CPAN::FirstTime defaults with nonexisting site dirs if a parent is writable
        DEBPKG:fixes/memoize_storable_nstore - [rt.cpan.org #77790] http://bugs.debian.org/587650 Memoize::Storable: respect 'nstore' option not respected
        DEBPKG:fixes/net_ftp_failed_command - [rt.cpan.org #37700] http://bugs.debian.org/491062 Net::FTP: cope gracefully with a failed command
        DEBPKG:fixes/perlbug-patchlist - [3541c11] http://bugs.debian.org/710842 [perl #118433] Make perlbug look up the list of local patches at run time
        DEBPKG:fixes/module_metadata_security_doc - [68cdd4b] CVE-2013-1437 documentation fix
        DEBPKG:fixes/module_metadata_taint_fix - [bff978f] http://bugs.debian.org/722210 [rt.cpan.org #88576] untaint version, if needed, in Module::Metadata
        DEBPKG:fixes/IPC-SysV-spelling - http://bugs.debian.org/730558 [rt.cpan.org #86736] Fix spelling of IPC_CREAT in IPC-SysV documentation
        DEBPKG:fixes/fix-undef-source -
  Built under linux
  Compiled at Mar 27 2014 18:30:28
  @INC:
    /etc/perl
    /usr/local/lib/perl/5.18.2
    /usr/local/share/perl/5.18.2
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.18
    /usr/share/perl/5.18
    /usr/local/lib/site_perl
    .

**Comment by oetiker** [comment]
my output looks like this :)

```
5392
INTERVAL
got HUP
INTERVAL
got HUP
got HUP
got HUP
INTERVAL
got HUP
got HUP
got HUP
got HUP
got HUP
```

I am also running  3.19.0-43-generic (ubuntu 14.04 and perl 5.18) ... 

mojolicious version 6.46

so what about this:

```perl
use Mojo::IOLoop; 
use Time::HiRes qw(gettimeofday);


print "$$\n";
$SIG{HUP} = sub { warn gettimeofday." got HUP\n" };
Mojo::IOLoop->recurring(5 => sub { warn gettimeofday." INTERVAL\n" });
# Mojo::IOLoop->recurring(0.1 => sub { });
Mojo::IOLoop->start;
```

does it make a difference if you remove the 0.1 line ?

**Comment by hadfl** [comment]
just tested on debian wheezy and omnios r16. can't reproduce the error.

```
$ uname -a
Linux nemesis 3.2.0-4-amd64 #1 SMP Debian 3.2.73-2+deb7u2 x86_64 GNU/Linux

$ perl -v
This is perl 5, version 14, subversion 2 (v5.14.2) built for x86_64-linux-gnu-thread-multi
(with 89 registered patches, see perl -V for more detail)

$ ./signaltest.pl 
6.48
20220
1456226382.99455 INTERVAL
1456226387.9948 INTERVAL
1456226389.92705 got HUP
1456226391.60293 got HUP
1456226392.61501 got HUP
1456226392.63094 got HUP
1456226392.98292 got HUP
1456226392.99508 INTERVAL
1456226393.33497 got HUP
1456226394.70292 got HUP
1456226395.37199 got HUP
1456226395.90299 got HUP
1456226396.49203 got HUP
1456226397.42295 got HUP
1456226397.89492 got HUP
1456226397.99609 INTERVAL
1456226398.33498 got HUP
1456226399.62302 got HUP
1456226402.99628 INTERVAL
```

```
$ uname -a
SunOS helios 5.11 omnios-33c53a8 i86pc i386 i86pc

$ perl -v
This is perl 5, version 16, subversion 1 (v5.16.1) built for i86pc-solaris-thread-multi-64int

$ ./signaltest.pl 
6.48
9533
1456227469.92335 INTERVAL
1456227474.93363 INTERVAL
1456227479.94388 INTERVAL
1456227481.72203 got HUP
1456227482.52791 got HUP
1456227483.25781 got HUP
1456227483.76168 got HUP
1456227484.40795 got HUP
1456227484.7058 got HUP
1456227484.94399 INTERVAL
1456227485.20988 got HUP
1456227486.66578 got HUP
1456227489.27309 got HUP
1456227489.94434 INTERVAL
1456227490.67406 got HUP
1456227491.65716 got HUP
```

**Comment by martin-rueegg** [comment]
thanks for your support!

i do have mojolicious version 6.4**3**

`signaltest.pl` (Version 1)

```perl
#!/usr/bin/perl
use Mojo::IOLoop;
use Mojolicious;
use Time::HiRes qw(gettimeofday);

print $Mojolicious::VERSION."\n";
print "$$\n";
$SIG{HUP} = sub { warn gettimeofday." got HUP\n" };
Mojo::IOLoop->recurring(5 => sub { warn gettimeofday." INTERVAL\n" });
# Mojo::IOLoop->recurring(0.5 => sub { });
Mojo::IOLoop->start;
```
result:
```
~# ./signaltest.pl
4.63
17927
1456229898.39448 INTERVAL
1456229903.39558 INTERVAL
1456229908.39469 INTERVAL
1456229913.39376 got HUP
1456229913.39381 INTERVAL
1456229918.39171 got HUP
1456229918.39176 INTERVAL
1456229923.39085 got HUP
1456229923.39089 INTERVAL
^C
```

`signaltest.pl` (Version 2: with second `recurring`)

```perl
#!/usr/bin/perl
use Mojo::IOLoop;
use Mojolicious;
use Time::HiRes qw(gettimeofday);

print $Mojolicious::VERSION."\n";
print "$$\n";
$SIG{HUP} = sub { warn gettimeofday." got HUP\n" };
Mojo::IOLoop->recurring(5 => sub { warn gettimeofday." INTERVAL\n" });
Mojo::IOLoop->recurring(0.5 => sub { });
Mojo::IOLoop->start;
```

result:
```
# ./signaltest.pl
4.63
17974
1456230250.60046 INTERVAL
1456230255.10007 got HUP
1456230255.59968 INTERVAL
1456230257.59961 got HUP
1456230258.10004 got HUP
1456230258.59998 got HUP
1456230259.10023 got HUP
1456230259.60024 got HUP
1456230260.09976 got HUP
1456230260.59986 got HUP
1456230260.59991 INTERVAL
1456230261.10035 got HUP
1456230261.60023 got HUP
1456230262.10029 got HUP
1456230262.5995 got HUP
1456230263.10018 got HUP
1456230263.5999 got HUP
1456230264.10012 got HUP
1456230264.60001 got HUP
1456230265.09976 got HUP
1456230265.59937 INTERVAL
^C
```

that's what is to be expected, as now the second `recurring` is fired every half a second.

this corresponds to my second approach: https://github.com/oetiker/znapzend/pull/191

can it be the mojolicious version? or the Mojo::IOLoop/Mojo version?

**Comment by oetiker** [comment]
oh yes ... you do have a rather old version ... are you NOT using the release tar to build your setup ?

**Comment by martin-rueegg** [comment]
yes, i did:
from `config.log`:
```
# more config.log
------8<--------------
It was created by znapzend configure 0.15.2, which was
generated by GNU Autoconf 2.69.  Invocation command line was
------8<--------------
configure:2901: checking for perl module 'Mojolicious@6.46'
configure:2908: result: Auto Install
configure:2901: checking for perl module 'Mojo::IOLoop::ForkCall@0.17'
configure:2908: result: Auto Install
configure:2901: checking for perl module 'Test::Harness'
configure:2908: result: Auto Install
configure:2901: checking for perl module 'TAP::Parser'
configure:2908: result: Auto Install
------8<--------------
```
**however**, i then installed the packages via `apt-get` :-)
`apt-get install libtest-harness-perl libmojolicious-perl`

so i have now removed the ubuntu package and re-configure'd & make'ed the znapzend again and it installed Mojolicious@6.46.

with Mojolicious v6.46 the problem's gone.

i'll update my PR https://github.com/oetiker/znapzend/pull/191 to only run the IOLoop if Mojolicious < v6.46.

**Comment by oetiker** [comment]
how did  you manage to setup without the correct modules getting installed ? did you just copy stuff from the source tree and circumvent the whole setup procedure ?

also note that the problem was that you got mojo 4.63 which is really really old.

**Comment by martin-rueegg** [comment]
Thank you, @oetiker and @hadfl for your support in investigating the problem. And my apologies for not having adhered to the standard installation procedure!
Do you think we should still do some sanity check during process start-up and either [warn] a deprecation or issue a [FATAL]?
Or shall I look into the fact, that `make` didn't check? (or I've not noticed the warning?)

@chriseqipe, did you follow the issue? What is your version of Mojolicious?
otherwise this issue can be closed as far as I'm concerned.
Thanks.

**Comment by oetiker** [comment]
the fact that you could run make and install and NOT get a copy of the packaged mojolicious installed, THAT is the interesting part ...

**Comment by martin-rueegg** [comment]
Ok, so i will have a look if I can reproduce that.

**Comment by martin-rueegg** [comment]
well, the problem is, that if the current ubuntu 14.04 package for mojolicious (`libmojolicious-perl:all 4.63+dfsg-1`) is installed, it takes precedence over the also `Successfully installed Mojolicious-6.46` from the `make install` call!

so somehow the library search path should be set to favour the local install of mojolicious ...

**Comment by oetiker** [comment]
have a look at the installed znapzend script ... at the very top should be a use lib statement to that effect.

**Comment by martin-rueegg** [comment]
i guess you intend the following lines in installed `/opt/znapzend-0.15.2/bin/znapzend` 
```perl
#!/usr/bin/perl

use lib qw(); # PERL5LIB
use lib qw(/opt/znapzend-0.15.2/lib); # LIBDIR

#...
```
apparently it is an other module that actually causes the issue! I have added a debug message, printing `@INC` and the loaded modules with the paths they are loaded from:
```perl
$self->zLog->debug('@INC='."qq(@INC)");
$self->zLog->debug($_ . " => " . $INC{$_}) for keys %INC;
```

it took me a while to find out on how to make the error reproducible! this due to the fact, that the installation of a ubuntu module might install some dependencies, which are not automatically uninstalled on package removal. but to cut the long story short, here is what i found

- the error can be reproduced
- it is not depending on `./configure`, `make` or even `make install`
- it is solely depending of **whether or not the ubuntu package `libev-perl` is installed!** <br /> (which can be found out by `dpkg -l libev-perl`)

the ubuntu package `libev-perl` installs the `/usr/lib/perl5/EV.pm` file (together with some dependencies). in my case, it is EV v4.15.

if this module is installed, Mojo::Reactor uses `Mojo::Reactor::EV` for it's loops, otherwise `Mojo::Reactor::Poll`. you can check which implementation is used by patching `lib/Mojo/Reactor.pm` as follows:

```
--- /tmp/znapzend-0.15.2/thirdparty/lib/perl5/Mojo/Reactor.pm   2016-02-12 23:23:54.000000000 +0100
+++ /opt/znapzend-0.15.2/lib/Mojo/Reactor.pm    2016-02-25 13:22:03.832767527 +0100
@@ -8,6 +8,9 @@

 sub detect {
   my $try = $ENV{MOJO_REACTOR} || 'Mojo::Reactor::EV';
+  my $using = load_class($try) ? 'Mojo::Reactor::Poll' : $try;
+  print "using $using\n";
+  return $using;
   return load_class($try) ? 'Mojo::Reactor::Poll' : $try;
 }

```

with ...
- `Mojo::Reactor::Poll` signals are received instantly
- `Mojo::Reactor::EV` only on the next timer

in theory, the `$ENV{MOJO_REACTOR}` could simply be set in `bin/znapzend` to always read `Mojo::Reactor::Poll` an thus to enforce this implementation, but according tot the Mojo::Reactor documentation, EV (event) is preferred over Poll:

> ### detect
>
>   `my $class = Mojo::Reactor->detect;`
> 
> Detect and load the best reactor implementation available, will try the value of the `MOJO_REACTOR` environment variable, `Mojo::Reactor::EV` or `Mojo::Reactor::Poll`.
> 
>   # Instantiate best reactor implementation available
>   `my $reactor = Mojo::Reactor->detect->new;`

i tried to determin which reactor is used, by checking `Mojo::Reactor->detect()`. however, this returns `Mojo::Reactor::EV` even if `/usr/lib/perl5/EV.pm` is not installed. so this is not reliable by itself.

so the following patch does activate the workaround we had elaborated only if required:
```
--- lib/ZnapZend.pm     2016-02-23 18:15:35.406203619 +0100
+++ /opt/znapzend-0.15.2/lib/ZnapZend.pm        2016-02-25 15:44:59.146920722 +0100
@@ -4,6 +4,7 @@
 use Mojo::IOLoop::ForkCall;
 use Mojo::Util qw(slurp);
 use Mojo::Log;
+use Mojo::Reactor;
 use ZnapZend::Config;
 use ZnapZend::ZFS;
 use ZnapZend::Time;
@@ -11,6 +12,7 @@
 use Scalar::Util qw(blessed);
 use Sys::Syslog;
 use File::Basename;
+use Class::Load qw(is_class_loaded);

 ### loglevels ###
 my %logLevels = (
@@ -527,6 +529,20 @@

     $self->zLog->info("znapzend (PID=$$) starting up ...");

+    # https://github.com/oetiker/znapzend/issues/168
+    my $reactor = Mojo::Reactor->detect();
+    my $is_ev = 0;
+    my $reactor_comment = ($reactor eq 'Mojo::Reactor::EV')
+                        ? '(' . (($is_ev = is_class_loaded('EV')) ? 'with' : 'without' ) . ' EV)'
+                        : '';
+    $self->zLog->debug("reactor implementation used by Mojo::Reactor: '$reactor' $reactor_comment");
+
+    unless ($reactor eq 'Mojo::Reactor::Poll' || !$is_ev)
+    {
+      $self->zLog->debug("Activating workaround to receive signals immediately");
+      Mojo::IOLoop->recurring(1 => sub { });
+    };
+
     $self->$daemonize if $self->daemonize;

     # set signal handlers
```

i found this solution, as the the following differences showed up in the loaded modules:

**with** `/usr/lib/perl5/EV.pm` installed:
```
[debug] EV.pm => /usr/lib/perl5/EV.pm
[debug] Mojo/Reactor/EV.pm => /opt/znapzend-0.15.2/lib/Mojo/Reactor/EV.pm
[debug] common/sense.pm => /usr/lib/perl5/common/sense.pm
```
**without**  `/usr/lib/perl5/EV.pm` installed:
```
[debug] Mojo/Reactor/EV.pm =>
```

i did successfully test the following conditions

`EV.pm` | `MOJO_REACTOR` | implementation used | workaround | signal reaction
------------ | ------------- | ------------- | ------------- | -------------
v4.15 | -- | 'Mojo::Reactor::EV' (with EV) | active | 1 sec
v4.15 | 'Mojo::Reactor::Poll' |  'Mojo::Reactor::Poll' | -- | immediate
v4.15 | 'Mojo::Reactor::EV' |  'Mojo::Reactor::EV' (with EV) | active | 1 sec
n/a | -- | 'Mojo::Reactor::EV' (without EV) | -- | immediate
n/a | 'Mojo::Reactor::Poll' | 'Mojo::Reactor::Poll' | -- | immediate
n/a | 'Mojo::Reactor::EV' | 'Mojo::Reactor::EV' (without EV) | -- | immediate

however i don't know how to write a test about that.

i'll update the PR.

**Comment by martin-rueegg** [comment]
oops, PR already merged. so i will make a new one then.

**Comment by oetiker** [comment]
no PR necessary ... c4f3f43 ... thanks!

**Comment by martin-rueegg** [comment]
well, i've already created the patch. have worked all day to figure out what the heck is going on.

so not fancy to take my version (d6cc401)?

**Comment by oetiker** [comment]
the figuring out is great :) thank you very much! 

we did see a similar patch create by sebastian in one of his mojo examples but there was no comment on why he did that ... making the whole thing conditional ... I don't think that this is necessary ... running this extra code will not have any noticable impact on proceedings.

**Comment by martin-rueegg** [comment]
i see. nevertheless quite frustrating.

**Comment by oetiker** [comment]
how about I edit the author of the commit to be you ?

**Comment by martin-rueegg** [comment]
thank you, that's kind of you. but don't bother.

**Comment by oetiker** [comment]
as you wish ... (the command is already sitting in my terminal, I just have to press return)

**Comment by martin-rueegg** [comment]
well, in that case `[Enter]` is only one key stroke, while `^C` are two.

so do whatever you like. - i learned a lot of perl today. be that my reward. :-)
